Add voucher tools to reverse and correct posted records

jniles opened the pull request:
Documented thoroughly in discussion #3063 .

Closes #3063.  Closes #3149.

sfount committed:
feat(vouchers): voucher-tools component structure

This commit introduces infrastructure for the `voucher-tools` collection
of components. `bh-voucher-tools-reverse` is introduced to reverse any
given generic transaction.

sfount committed:
feat(vouchers): linking reversal tool utilities

This commit correctly implements and links the utilities for the
`bh-voucher-tools-reverse` component, the journal/reverse route was
updated to provide all information on the voucher that is generated
through the reversal process.

It also adds English translation for the afformentioned component.

sfount committed:
chore(vouchers): cleanup voucher tool markup, lint

This commit fully integrates the `bh-voucher-tools-reverse` into the
journal transaction editing modal.
It also
* Updates the sytle to clearly indicate this is a seperate tool
* Badge style
* Remove unused markup

sfount committed:
feat(vouchers): server voucher tools API end point

This commit introduces an API end point for the `correct` voucher tool.
This method is responsible for both reversing any journal transaction
and creating a generic voucher to replace it with corrected values. This
is achieved by calling the server controllers responsible for journal
transaction reversal and voucher creation and ensuring that the results
are cleaned up on failure.

sfount committed:
feat(vouchers): integrate component, service, API

This commit introduces the core files needed for voucher tools on the
client, the `voucher-tools-correct` component, VoucherToolsService and
templates. It also hooks all of the pieces together to pass information
on the currently edited transaction through to the server correct
transaction API end point.

sfount committed:
feat(vouchers): voucher tools correct component

This commit completes the intial API and client integration for the
voucher tools correct component, including integration with
`bh-voucher-tools-badge` and the additional `correcting` state on
`editTransaction` modal.

sfount committed:
chore(vouchers): correct component style standards

This commit updates the voucher tools correct component to fit the
voucher tools standard. It also cleans up controllers and the Server
API
cleanup endpoint to include Entity `reservsed` flag undoing.

sfount committed:
chore(vouchers): integrate editmodal correct style

This commit update the `editTransaction` modal to properly integrate the
`voucher-tools` `correct` component. This tool differs from reverse tool
as it is not self contained, it requires the containing controller to
provide edited information thorugh the `source` binding. This commit
updates the edit modal to allow the user to make 'corrections' to the
transaction before submitting the correct tool with the latest changes.

sfount committed:
chore(vouchers): clean up development structures

This commit cleans up the controller code for voucher tools components
and updates comments and documentation to ensure the PR is up to project
standards.

sfount committed:
chore(vouchers): cleanup and update structure

This commit removes out of date comments and any additional development
code that should not be merged into production.

jniles committed:
chore(journal): add French i18n translations

This commit adds the French i18n for @sfount's correction tool.

jniles committed:
chore: add SP to migrate.sql

This commit adds the stored procedure for undoing reversals to
migrate.sql.

jniles committed:
test(journal): integration test for reversals

This commit adds an integration test for the journal corrections.  It
also caught a bug in the cash data SQL that led to an unbalanced
transaction.

jniles committed:
test(vouchers): integration test corrections

This commit adds an integration test for the corrections voucher tool.

mbayopanda commented on client/src/css/structure.css:
```diff
@@ -926,3 +926,46 @@ a[disabled] {
+
+.voucher-tools-underline {
+  margin-top: 2px;
+  margin-bottom: 10px;
```
css best practice : Rule must have all its properties in alphabetical order

jniles replied:
I like it. :+1:  Will do.

mbayopanda commented on client/src/css/structure.css:
```diff
@@ -926,3 +926,46 @@ a[disabled] {
+
+.voucher-tools-header-element {
+  /* display : inline-block; */
+}
```
This rule is empty, is it possible to remove it

mbayopanda commented on client/src/css/structure.css:
```diff
@@ -926,3 +926,46 @@ a[disabled] {
+.voucher-tools-container {
+  border : 1px solid rgb(46, 109, 164);
+  border : 1px solid rgba(46, 109, 164, 0.8);
+  box-shadow:0px 1px 1px #bdbdbd;
```
css best practice : Values of 0 shouldn't have units specified

mbayopanda commented on server/controllers/finance/journal/index.js:
```diff
@@ -571,12 +573,25 @@ function transformColumns(rows, newRecord, transactionToEdit, setFiscalData) {
+
+  reverseTransaction(recordUuid, req.session.user.id, req.body.description)
+    .then((reverseResult) => VoucherService.lookupVoucher(reverseResult.uuid))
+    .then((voucher) => res.status(201).json({ uuid : voucher.uuid, voucher }))
```
Parenthesis are not necessary for one argument : 
```js
.then(reverseResult => VoucherService.lookupVoucher(reverseResult.uuid))
.then(voucher => res.status(201).json({ uuid : voucher.uuid, voucher }))
```

mbayopanda reviewed (commented):
I made a first review and i just notice that when we correct a transaction with the voucher-tool, the transaction is submitted without values validation. When i set debit = 77 and credit = 75, i got : 

![image](https://user-images.githubusercontent.com/5445251/46732097-7a099d80-cc84-11e8-97ad-de91be3039ab.png)

jniles committed:
fix(reversal): undo reversal failures

We needed to wrap the ReverseTransaction() logic in a db.transaction()
to ensure that it was cleaning up its values.

jniles committed:
feat(editTransaction): move offline to service

This commit moves the offline validation into the TransactionService.
The motivation is to have better separation of concerns and to share the
offline validation between the voucher tools and the main edit
transaction modal.

jniles committed:
test(transaction): add tests for offlineValidation

This commit introduces unit tests for the offlineValidation functions of
the TransactionService.

jniles committed:
fix(vouchers): update tests for latest changes

This commit updates the voucher tests to accounts for the latest changes
in the module.

mbayopanda commented on client/src/js/services/TransactionService.js:
```diff
@@ -71,5 +74,77 @@ function TransactionService($http, util, Modal) {
+    const hasImbalancedTransaction = Number(debits.toFixed('2')) !== Number(credits.toFixed('2'));
+    if (hasImbalancedTransaction) {
+      return ERROR_IMBALANCED_TRANSACTION;
+    }
```
This code looks good but i noticed that the validation is skipped for unbalanced amounts : 

![2018-11-09_10-38-15](https://user-images.githubusercontent.com/5445251/48255025-d68ae480-e40b-11e8-8109-626afd9a9797.gif)

jniles replied:
Hmm... I can't reproduce this on chrome or firefox.  Let me see if I can reproduce it on a VM.

mbayopanda commented on server/controllers/finance/journal/index.js:
```diff
@@ -595,34 +611,43 @@ function reverse(req, res, next) {
+      'The transaction has been already cancelled',
+      'POSTING_JOURNAL.ERRORS.MULTIPLE_CANCELLING'
+    );
+  }
```
👍

mbayopanda commented on client/src/modules/voucher-tools/correct/voucher-tools.correct.component.js:
```diff
@@ -0,0 +1,103 @@
+function bhVoucherToolsCorrect(VoucherTools, Transactions, $translate) {
+  const $ctrl = this;
+  const VOUCHER_TOOLS_REVERSE_DESCRIPTION = 'VOUCHERS.TOOLS.REVERSE.DESCRIPTION';
+  const VOUCHER_TOOLS_CORRECTION_DESCRIPTION = 'VOUCHERS.TOOLS.CORRECT.DESCRIPTION';
```
The entry `VOUCHERS.TOOLS.CORRECT.DESCRIPTION` is missing in the voucher.json file.

mbayopanda reviewed (approved):
This is a great work, well done

I made some comments but the main feature `correct` and `reverse` transaction works perfectly.
